_posts/2020-07-08-event-driven-functions-with-openfaas-and-nats.md

```diff
@@ -0,0 +1,146 @@
+
+* [k3d](https://github.com/rancher/k3d) to create a [k3s](https://github.com/rancher/k3s) cluster. k3s is lightweight distribution of kubernetes from Rancher Labs.
+* `kubectl` to manage kubernetes cluster.
+* `helm` to install `nats-connector` using helm chart
+* [arkade](https://github.com/alexellis/arkade) to install OpenFaaS. `arkade` is simple CLI tool to install helm charts and apps to your cluster in one command.
+* [faas-cli](https://github.com/openfaas/faas-cli) CLI tool to manage OpenFaaS functions.
+
+> Please make sure you have these tools installed before you proceed next.
+
+## Install Components
+
+#### Create a kubernetes cluster
+
+`k3d` is a tool to run `k3s` in docker. Create a single node `k3s` cluster.
+
+```
+k3d create
+```
+
+Follow instructions returned in the output of this command to configure KUBECONFIG to access the cluster using kubectl.
+
+#### Install OpenFaaS 
+Install OpenFaaS using `arkade`
+
+```
+arkade install openfaas --set basic_auth=true --set functionNamespace=openfaas-fn
+```
+
+Follow instructions printed to obtain the gateway password and login with `faas-cli`
+
+#### Install NATS connector using helm chart
+`nats-connector` is connector which invokes OpenFaaS function when a message is sent to a NATS topic.
+
+```
+helm repo add nats-connector  https://openfaas.github.io/faas-netes
+
+helm install nats-connector openfaas/nats-connector --namespace=openfaas
+
+```
+
+> NATS comes with OpenFaaS installation, so we are not setting up NATS here
+
+#### Login to OpenFaaS gateway using CLI
+
+Port forward the gateway service to access it. 
+
+```
+kubectl port-forward -n openfaas svc/gateway 8080:8080 &
+```
+
+Get the password
+
+```
+PASSWORD=$(kubectl get secret -n openfaas basic-auth -o jsonpath="{.data.basic-auth-password}" | base64 --decode; echo)
+```
+
+Login to the gateway
+
+```
+echo -n $PASSWORD | faas-cli login --username admin --password-stdin
+```
+
+
+## Deploy functions
+`nats-connector` repo comes with test functions to verify installation. It has two OpenFaas functions.
+
+`receive-message` function subscribes to NATS topic nats-test. On every new message on nats-test topic, receive-message will be invoked.
+
+`publish-message` function publishes message to nats-test topic.
+
+Deploy `receive-message` function
+
+```
+faas-cli deploy --name receive-message --image openfaas/receive-message:latest --fprocess='./handler' --annotation topic="nats-test"
+```
+
+Deploy `publish-message` function
+```
+faas-cli deploy --name publish-message  --image openfaas/publish-message:latest  --fprocess='./handler' --env nats_url=nats://nats.openfaas:4222
+```
+
+> Note: You can also build and deploy this function using the stack.yml and code present in nats-connector repository. But for simplicity, I am using pre built and published images of  these functions.
+
+Invoke `publish-message` function to publish a test message
+
+```
+faas-cli invoke publish-message <<< "test message"
```

Comment:
I think Alex prefers and it would be more consistent with previous posts to use echo

```suggestion
echo "test message" | faas-cli invoke publish-message
```

Reply:
Updated